### ISSUE TITLE
IngestExternalFile and Compaction caused deadlock bug

### ISSUE THREAD

**Issue by gdyang1990** [facebook/rocksdb#5007]
We use the IngestExternalFile interface to load a bunch of data into RocksDB. And we found that in some scenarios this call will always block.

The release version we use is 5.17.2

Through pstack analysis, it seems that RocksDB seems to be deadlocked. A few key call stacks are as follows:

IngestExternalFile stack
```
Thread 144 (Thread 0x7fe5015fa700 (LWP 515446)):
#0  0x00007fe680f423cc in pthread_cond_wait@@GLIBC_2.3.2 () from /opt/compiler/gcc-4.8.2/lib64/libpthread.so.0
#1  0x0000000001fb2647 in rocksdb::port::CondVar::Wait() ()
#2  0x0000000001edbc49 in rocksdb::InstrumentedCondVar::Wait() ()
#3  0x0000000002014f0d in rocksdb::DBImpl::WaitUntilFlushWouldNotStallWrites(rocksdb::ColumnFamilyData*, bool*) ()
#4  0x00000000020196b6 in rocksdb::DBImpl::FlushMemTable(rocksdb::ColumnFamilyData*, rocksdb::FlushOptions const&, rocksdb::FlushReason, bool) ()
#5  0x0000000002000bdf in rocksdb::DBImpl::IngestExternalFile(rocksdb::ColumnFamilyHandle*, std::vector<std::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&, rocksdb::IngestExternalFileOptions const&) ()
```

BackgroundCompaction stack
````
Thread 156 (Thread 0x7fe513bff700 (LWP 515434)):
#0  0x00007fe680f423cc in pthread_cond_wait@@GLIBC_2.3.2 () from /opt/compiler/gcc-4.8.2/lib64/libpthread.so.0
#1  0x0000000001fb2647 in rocksdb::port::CondVar::Wait() ()
#2  0x0000000001edbc49 in rocksdb::InstrumentedCondVar::Wait() ()
#3  0x0000000001fefe72 in rocksdb::DBImpl::WaitForIngestFile() ()
#4  0x0000000002020cfe in rocksdb::DBImpl::BackgroundCallCompaction(rocksdb::DBImpl::PrepickedCompaction*, rocksdb::Env::Priority) ()
#5  0x0000000002021154 in rocksdb::DBImpl::BGWorkCompaction(void*) ()
#6  0x00000000020dc9ff in std::_Function_handler<void ()(), std::_Bind<void (*()(void*))(void*)> >::_M_invoke(std::_Any_data const&) ()
#7  0x0000000002089d66 in std::function<void ()()>::operator()() const ()
#8  0x00000000020db687 in rocksdb::ThreadPoolImpl::Impl::BGThread(unsigned long) ()
#9  0x00000000020db737 in rocksdb::ThreadPoolImpl::Impl::BGThreadWrapper(void*) ()
#10 0x00000000020dca10 in _ZNSt6thread11_State_implINS_8_InvokerISt5tupleIJPFPvS3_EPN7rocksdb16BGThreadMetadataEEEEEE6_M_runEv ()
#11 0x000000000271974f in execute_native_thread_routine ()
#12 0x00007fe680f3e1c3 in start_thread () from /opt/compiler/gcc-4.8.2/lib64/libpthread.so.0
#13 0x00007fe68123b12d in clone () from /opt/compiler/gcc-4.8.2/lib64/libc.so.6
````

By reading the source code, IngestExternalFile sometimes requires a FlushMemTable with option allow_write_stall=false.
And BackgroundCompaction will check if IngestFile is happening. 
So they form waiting for each other, which leads to deadlock.

**Comment by riversand963** [comment]
Thanks @gdyang1990 for reporting. I believe this should have been fixed by 9be3e6b48884c80733fa791b982a02f62a199e92. Can you confirm your RocksDB version?

**Comment by gdyang1990** [comment]
> Thanks @gdyang1990 for reporting. I believe this should have been fixed by [9be3e6b](https://github.com/facebook/rocksdb/commit/9be3e6b48884c80733fa791b982a02f62a199e92). Can you confirm your RocksDB version?
Thank you for your reply.
Yes. It has been fixed.  The version we used was 5.17.2, and the bug was fixed after this version.
